### ISSUE TITLE
Handling Anonymous User (not 'SimpleLazyObject' errors)

### ISSUE THREAD

**Issue by danielsokolowski** [django-guardian/django-guardian#152]
Need to consensus if this is should be addressed in django-guardian or on the project level individually or even Django itself  - so do please speak you mind.

The built in django contrib auth returns a custom AnonymousUser which is of course not the same as the django guardian concept of anonymous user set through ANONYMOUS_USER_ID. 

This means  django-guardian will error if `request.user`  is not authenticated for example when using the `get_objects_for_user(user=request.user, ...)` shortcut:

```
TypeError at /offers/negotiate/3/
int() argument must be a string or a number, not 'SimpleLazyObject'
```

There are three solutions that come to mind, 

1. Modify shortcut code to check if user.is_anonymous and if so use guardian's anonymous user in lieu.
2. Create a middleware that would replace `request.user` with the guardian anonymous user instance as needed.
3. Modify Django to provide an ability to specify a custom AnonymousUser class

The more I think about the more I think option number 1 is the best approach - there is ofcourse the option of doing nothing and prior to using the shorctuts casting as needed to the guradians AnonymousUser instance.

Thoughts?

**Comment by maenu** [comment]
Took me a while to find this out too. I added a middleware to make it work:

```
from guardian.utils import get_anonymous_user


class GuardianAuthenticationMiddleware(object):

    def process_request(self, request):
        if request.user.is_anonymous():
            request.user = get_anonymous_user()
```

**Comment by brianmay** [comment]
Is this still the case? Unfortunately you didn't provide the full stack trace, this makes it hard to understand what was going wrong. Or if it has been fixed.

**Comment by brianmay** [comment]
Looks like this was already fixed in 960bf7ba165cfd437d40107dca31f7ec85ea4405.
